### PR TITLE
feat(query): add udf priv

### PR DESCRIPTION
**Pull request by TCeason**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support grant udf privilege.

Now usage has `usage` privilege.

| Privileges | Usage |
|--|--|
| usage | Enables calling a UDF |
```sql
GRANT [ ALL | USAGE ] ON UDF <udf_name> TO <user_name>
GRANT [ ALL | USAGE ] ON UDF <udf_name> TO <role_name>
```

`all` is equal with `usage`.

## Note

1. don't need to check the udf auth if it's already be constantly folded.
2. don't need to check the udf auth if it's a value in insert.

- Closes #https://github.com/datafuselabs/databend/issues/13318

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13735)

### CODE REVIEW

File: tests/suites/0_stateless/18_rbac/18_0001_udf_priv.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+echo "drop user if exists 'test-user'" | $BENDSQL_CLIENT_CONNECT
+
+## create user
+echo "create user 'test-user' IDENTIFIED BY '$TEST_USER_PASSWORD'" | $BENDSQL_CLIENT_CONNECT
+
+
+echo "DROP FUNCTION IF EXISTS test_alter_udf;" |  $BENDSQL_CLIENT_CONNECT
+echo "CREATE FUNCTION test_alter_udf AS (p) -> not(is_null(p))" | $BENDSQL_CLIENT_CONNECT
+#error test
+echo "select test_alter_udf(1)" | $TEST_USER_CONNECT
```
Comment (TCeason) [inline tests/suites/0_stateless/18_rbac/18_0001_udf_priv.sh:17]: Need more check to cover the modify.

File: tests/suites/0_stateless/18_rbac/18_0001_udf_priv.sh [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+echo "grant insert, delete, update, select on default.t to 'test-user';" |  $BENDSQL_CLIENT_CONNECT
+echo "grant select on default.t to 'test-user';" |  $BENDSQL_CLIENT_CONNECT
+echo "grant super on *.* to 'test-user';" |  $BENDSQL_CLIENT_CONNECT
+sleep 2;
+
+# error test need privielge f2
+echo "=== No Privilege on f1, f2 ==="
+echo "select f2(f1(1));" | $TEST_USER_CONNECT
+# bug need error
+echo "select add(1, f2(f1(1)));" | $TEST_USER_CONNECT
```
Comment (TCeason) [inline tests/suites/0_stateless/18_rbac/18_0001_udf_priv.sh:30]: Udf in the query plan into a constant expr or string:

```

select add(1, f2(f1(1)));

plan is Query { s_expr: SExpr { plan: EvalScalar(EvalScalar { items: [ScalarItem { scalar: ConstantExpr(ConstantExpr { span: Some(7..24), value: Number(2_u8) }), index: 1 }] }), children: [SExpr { plan: Scan(Scan { table_index: 0, columns: {0}, push_down_predicates: None, limit: None, order_by: None, prewhere: None, agg_index: None, statistics: Statistics { statistics: None, col_stats: {0: None} } }), children: [], original_group: None, rel_prop: Mutex { data: None, poisoned: false, .. }, stat_info: Mutex { data: None, poisoned: false, .. }, applied_rules: AppliedRules { rules: RuleSet { rules: RoaringBitmap<[]> } } }], original_group: None, rel_prop: Mutex { data: None, poisoned: false, .. }, stat_info: Mutex { data: None, poisoned: false, .. }, applied_rules: AppliedRules { rules: RuleSet { rules: RoaringBitmap<[18]> } } }, metadata: RwLock { data: Metadata { tables: [TableEntry { catalog: "default", database: "system", name: "one", index: 0, .. }], columns: [BaseTableColumn(BaseTableColumn { table_index: 0, column_index: 0, column_name: "dummy", column_position: Some(1), data_type: Number(UInt8), path_indices: None, leaf_index: Some(0), virtual_computed_expr: None }), DerivedColumn(DerivedColumn { column_index: 1, alias: "add(1, f2(f1(1)))", data_type: Number(UInt8) })], lazy_columns: {}, table_row_id_index: {}, agg_indexes: {}, max_column_position: 0 } }, bind_context: BindContext { parent: Some(BindContext { parent: None, columns: [], bound_internal_columns: {}, aggregate_info: AggregateInfo { aggregate_functions: [], aggregate_arguments: [], group_items: [], output_columns: [], aggregate_functions_map: {}, group_items_map: {}, grouping_sets: None }, windows: WindowInfo { window_functions: [], window_functions_map: {} }, lambda_info: LambdaInfo { lambda_arguments: [], lambda_functions: [], lambda_functions_map: {} }, cte_name: None, cte_map_ref: {}, in_grouping: false, view_info: None, srfs: {}, expr_context: Unknown, allow_internal_columns: true, planning_agg_index: false, window_definitions: {} }), columns: [ColumnBinding { database_name: None, table_name: None, column_position: None, table_index: None, column_name: "add(1, f2(f1(1)))", index: 1, data_type: Number(UInt8), visibility: Visible, virtual_computed_expr: None }], bound_internal_columns: {}, aggregate_info: AggregateInfo { aggregate_functions: [], aggregate_arguments: [], group_items: [], output_columns: [], aggregate_functions_map: {}, group_items_map: {}, grouping_sets: None }, windows: WindowInfo { window_functions: [], window_functions_map: {} }, lambda_info: LambdaInfo { lambda_arguments: [], lambda_functions: [], lambda_functions_map: {} }, cte_name: None, cte_map_ref: {}, in_grouping: false, view_info: None, srfs: {}, expr_context: Unknown, allow_internal_columns: true, planning_agg_index: false, window_definitions: {} }, rewrite_kind: None, formatted_ast: None, ignore_result: false }


SELECT i,nth_value(i, f2(f1(2))) OVER (PARTITION BY i) fv FROM t;

plan is Query { s_expr: SExpr { plan: EvalScalar(EvalScalar { items: [ScalarItem { scalar: BoundColumnRef(BoundColumnRef { span: Some(7..8), column: ColumnBinding { database_name: Some("default"), table_name: Some("t"), column_position: Some(1), table_index: Some(0), column_name: "i", index: 0, data_type: Number(UInt8), visibility: Visible, virtual_computed_expr: None } }), index: 0 }, ScalarItem { scalar: BoundColumnRef(BoundColumnRef { span: None, column: ColumnBinding { database_name: None, table_name: None, column_position: None, table_index: None, column_name: "nth_value(i, f2(f1(2))) OVER (PARTITION BY i)", index: 1, data_type: Nullable(Number(UInt8)), visibility: Visible, virtual_computed_expr: None } }), index: 2 }] }), children: [SExpr { plan: Window(Window { span: Some(9..54), index: 1, function: NthValue(NthValueFunction { n: Some(2), arg: BoundColumnRef(BoundColumnRef { span: Some(19..20), column: ColumnBinding { database_name: Some("default"), table_name: Some("t"), column_position: Some(1), table_index: Some(0), column_name: "i", index: 0, data_type: Number(UInt8), visibility: Visible, virtual_computed_expr: None } }), return_type: Nullable(Number(UInt8)) }), arguments: [ScalarItem { scalar: BoundColumnRef(BoundColumnRef { span: Some(19..20), column: ColumnBinding { database_name: Some("default"), table_name: Some("t"), column_position: Some(1), table_index: Some(0), column_name: "i", index: 0, data_type: Number(UInt8), visibility: Visible, virtual_computed_expr: None } }), index: 0 }], partition_by: [ScalarItem { scalar: BoundColumnRef(BoundColumnRef { span: Some(52..53), column: ColumnBinding { database_name: Some("default"), table_name: Some("t"), column_position: Some(1), table_index: Some(0), column_name: "i", index: 0, data_type: Number(UInt8), visibility: Visible, virtual_computed_expr: None } }), index: 0 }], order_by: [], frame: WindowFuncFrame { units: Range, start_bound: Preceding(None), end_bound: Following(None) } }), children: [SExpr { plan: Scan(Scan { table_index: 0, columns: {0}, push_down_predicates: None, limit: None, order_by: None, prewhere: None, agg_index: None, statistics: Statistics { statistics: Some(TableStatistics { num_rows: Some(0), data_size: Some(0), data_size_compressed: Some(0), index_size: Some(0), number_of_blocks: None, number_of_segments: None }), col_stats: {0: None} } }), children: [], original_group: None, rel_prop: Mutex { data: None, poisoned: false, .. }, stat_info: Mutex { data: None, poisoned: false, .. }, applied_rules: AppliedRules { rules: RuleSet { rules: RoaringBitmap<[]> } } }], original_group: None, rel_prop: Mutex { data: None, poisoned: false, .. }, stat_info: Mutex { data: None, poisoned: false, .. }, applied_rules: AppliedRules { rules: RuleSet { rules: RoaringBitmap<[]> } } }], original_group: None, rel_prop: Mutex { data: None, poisoned: false, .. }, stat_info: Mutex { data: None, poisoned: false, .. }, applied_rules: AppliedRules { rules: RuleSet { rules: RoaringBitmap<[18]> } } }, metadata: RwLock { data: Metadata { tables: [TableEntry { catalog: "default", database: "default", name: "t", index: 0, .. }], columns: [BaseTableColumn(BaseTableColumn { table_index: 0, column_index: 0, column_name: "i", column_position: Some(1), data_type: Number(UInt8), path_indices: None, leaf_index: Some(0), virtual_computed_expr: None }), DerivedColumn(DerivedColumn { column_index: 1, alias: "nth_value(i, f2(f1(2))) OVER (PARTITION BY i)", data_type: Nullable(Number(UInt8)) }), DerivedColumn(DerivedColumn { column_index: 2, alias: "fv", data_type: Nullable(Number(UInt8)) })], lazy_columns: {}, table_row_id_index: {}, agg_indexes: {}, max_column_position: 0 } }, bind_context: BindContext { parent: Some(BindContext { parent: None, columns: [], bound_internal_columns: {}, aggregate_info: AggregateInfo { aggregate_functions: [], aggregate_arguments: [], group_items: [], output_columns: [], aggregate_functions_map: {}, group_items_map: {}, grouping_sets: None }, windows: WindowInfo { window_functions: [], window_functions_map: {} }, lambda_info: LambdaInfo { lambda_arguments: [], lambda_functions: [], lambda_functions_map: {} }, cte_name: None, cte_map_ref: {}, in_grouping: false, view_info: None, srfs: {}, expr_context: Unknown, allow_internal_columns: true, planning_agg_index: false, window_definitions: {} }), columns: [ColumnBinding { database_name: Some("default"), table_name: Some("t"), column_position: Some(1), table_index: Some(0), column_name: "i", index: 0, data_type: Number(UInt8), visibility: Visible, virtual_computed_expr: None }, ColumnBinding { database_name: None, table_name: None, column_position: None, table_index: None, column_name: "fv", index: 2, data_type: Nullable(Number(UInt8)), visibility: Visible, virtual_computed_expr: None }], bound_internal_columns: {}, aggregate_info: AggregateInfo { aggregate_functions: [], aggregate_arguments: [], group_items: [], output_columns: [], aggregate_functions_map: {}, group_items_map: {}, grouping_sets: None }, windows: WindowInfo { window_functions: [], window_functions_map: {} }, lambda_info: LambdaInfo { lambda_arguments: [], lambda_functions: [], lambda_functions_map: {} }, cte_name: None, cte_map_ref: {}, in_grouping: false, view_info: None, srfs: {}, expr_context: Unknown, allow_internal_columns: true, planning_agg_index: false, window_definitions: {} }, rewrite_kind: None, formatted_ast: None, ignore_result: false }

insert into t values (f2(f1(1)));

plan is Insert { catalog: "default", database: "default", table: "t", table_id: 294593, schema: TableSchema { fields: [TableField { name: "i", default_expr: None, data_type: Number(UInt8), column_id: 0, computed_expr: None }], metadata: {}, next_column_id: 1 }, overwrite: false }
source is data: "(f2(f1(1)))", start: 21

```

Comment (TCeason) [reply tests/suites/0_stateless/18_rbac/18_0001_udf_priv.sh:30]: cc @andylokandy @sundy-li 

For the insert, `Maybe we can forbidden udf in insert source?`

Constant can not limit in plan.

Comment (sundy-li) [reply tests/suites/0_stateless/18_rbac/18_0001_udf_priv.sh:34]: We don't need to check the udf auth if it's already be constantly folded.

**Comment by TCeason** [comment]
ut failed, but it wasn't relevant to this commit.

```
--- STDERR:              common-sql::location test_parse_uri_location ---
thread 'test_parse_uri_location' panicked at src/query/sql/tests/location.rs:422:9:
assertion `left == right` failed: s3_with_access_key_id
  left: (S3(StorageS3Config { endpoint_url: "https://s3.amazonaws.com/", region: "us-east-2", bucket: "test", root: "/tmp/", disable_credential_loader: true, enable_virtual_host_style: false, role_arn: "", external_id: "", access_key_id: "******_id", secret_access_key: "******key", security_token: "******ken", master_key: "", allow_anonymous: false }), "/")
 right: (S3(StorageS3Config { endpoint_url: "https://s3.amazonaws.com/", region: "ap-south-1", bucket: "test", root: "/tmp/", disable_credential_loader: true, enable_virtual_host_style: false, role_arn: "", external_id: "", access_key_id: "******_id", secret_access_key: "******key", security_token: "******ken", master_key: "", allow_anonymous: false }), "/")
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Review by xudong963 — APPROVED** [review]
binder and planner parts: LGTM

File: src/query/ast/src/parser/statement.rs [hunk 3]
```diff
@@ -2102,26 +2102,25 @@ pub fn grant_source(i: Input) -> IResult<AccountMgrSource> {
...
-    // ALL ~ PRIVILEGES? ~ ON ~ UDF ~ #ident
-    // },
-    // |(_, _, _, _, udf)| AccountMgrSource::Privs {
-    // privileges: vec![UserPrivilegeType::UsageUDF],
-    // level: AccountMgrLevel::UDF(udf.to_string()),
-    // },
-    // );
+    let udf_privs = map(
+        rule! {
+        USAGE ~ ON ~ UDF ~ #ident
```
Comment (andylokandy) [inline src/query/ast/src/parser/statement.rs:2107]:
```suggestion
            USAGE ~ ON ~ UDF ~ #ident
```

File: src/query/ast/src/parser/statement.rs [hunk 4]
```diff
@@ -2102,26 +2102,25 @@ pub fn grant_source(i: Input) -> IResult<AccountMgrSource> {
...
+        },
+        |(_, _, _, udf)| AccountMgrSource::Privs {
+            privileges: vec![UserPrivilegeType::Usage],
+            level: AccountMgrLevel::UDF(udf.to_string()),
+        },
+    );
+
+    let udf_all_privs = map(
+        rule! {
+        ALL ~ PRIVILEGES? ~ ON ~ UDF ~ #ident
```
Comment (andylokandy) [inline src/query/ast/src/parser/statement.rs:2117]:
```suggestion
            ALL ~ PRIVILEGES? ~ ON ~ UDF ~ #ident
```

File: src/query/ast/tests/it/testdata/statement.txt [hunk 5]
```diff
@@ -11306,6 +11306,131 @@ Grant(
 )
 
 
+---------- Input ----------
+REVOKE write ON stage s1 FROM a;
+---------- Output ---------
+REVOKE Write ON STAGE s1 FROM USER 'a'@'%'
```
Comment (andylokandy) [inline src/query/ast/tests/it/testdata/statement.txt:11312]: There is an additional `USER`. Is it expected?

File: src/query/sql/src/planner/optimizer/s_expr.rs [hunk 6]
```diff
@@ -320,5 +502,99 @@ fn find_subquery_in_expr(expr: &ScalarExpr) -> bool {
         ScalarExpr::CastExpr(expr) => find_subquery_in_expr(&expr.argument),
         ScalarExpr::SubqueryExpr(_) => true,
         ScalarExpr::UDFServerCall(expr) => expr.arguments.iter().any(find_subquery_in_expr),
+        ScalarExpr::UDFLambdaCall(expr) => find_subquery_in_expr(&expr.scalar),
+    }
+}
+
+pub fn get_udf_names(scalar: &ScalarExpr) -> Result<HashSet<String>> {
```
Comment (andylokandy) [inline src/query/sql/src/planner/optimizer/s_expr.rs:509]: Recommends to use `Visitor` for `ScalarExpr`. Here is an [example](https://github.com/datafuselabs/databend/blob/3f4e629f8ba38b030b6a9fa3fc84def4f90a61e9/src/query/sql/src/planner/binder/delete.rs#L187C1-L199)

File: src/query/ast/tests/it/testdata/statement.txt (hunk 5 above)
Comment (TCeason) [reply src/query/ast/tests/it/testdata/statement.txt:11312]: except. default is user, also can be role.

**Review by flaneur2020 — APPROVED** [review]

File: src/query/sql/src/planner/optimizer/s_expr.rs (hunk 6 above)
Comment (TCeason) [reply src/query/sql/src/planner/optimizer/s_expr.rs]: Fixed. Please review the newest commit.

File: src/query/sql/src/planner/optimizer/s_expr.rs [hunk 7]
```diff
@@ -506,95 +507,31 @@ fn find_subquery_in_expr(expr: &ScalarExpr) -> bool {
...
+            let name = &udf.func_name;
+            self.visit(&udf.scalar)?;
+            self.udfs.insert(name);
+            Ok(())
         }
     }
-    Ok(udfs)
+
+    let udfs = HashSet::new();
+    let mut find_udfs = FindUdfNames { udfs };
```
Comment (andylokandy) [inline src/query/sql/src/planner/optimizer/s_expr.rs:534]:
```suggestion
let mut find_udfs = FindUdfNames { udfs: HashSet::new() };
```

File: src/query/sql/src/planner/optimizer/s_expr.rs [hunk 8]
```diff
@@ -506,95 +507,31 @@ fn find_subquery_in_expr(expr: &ScalarExpr) -> bool {
...
-            ScalarExpr::BoundColumnRef(_) | ScalarExpr::ConstantExpr(_) => {}
+
+            self.udfs.insert(&udf.func_name);
+            Ok(())
+        }
+
+        fn visit_udf_lambda_call(&mut self, udf: &'a UDFLambdaCall) -> Result<()> {
+            let name = &udf.func_name;
+            self.visit(&udf.scalar)?;
+            self.udfs.insert(name);
```
Comment (andylokandy) [inline src/query/sql/src/planner/optimizer/s_expr.rs:528]:
```suggestion
self.visit(&udf.scalar)?;
self.udfs.insert(&udf.func_name);
```

**Review by andylokandy — APPROVED** [review]

File: src/query/sql/src/planner/optimizer/s_expr.rs (hunk 7 above)
Comment (TCeason) [reply src/query/sql/src/planner/optimizer/s_expr.rs]: Fixed.

File: src/query/sql/src/planner/optimizer/s_expr.rs (hunk 8 above)
Comment (TCeason) [reply src/query/sql/src/planner/optimizer/s_expr.rs]: Fixed.

**Pull request merged**